dbms/src/Functions/FunctionsJson.h

```diff
@@ -530,12 +607,658 @@ class FunctionsJsonArray : public IFunction
             offsets_to[i] = write_buffer.count();
             for (const auto & source : sources)
             {
-                if (source)
+                if constexpr (is_input_nullable)
+                {
+                    if (source)
+                        source->next();
+                }
+                else
+                {
+                    assert(source);
                     source->next();
+                }
+            }
+        }
+    }
+};
+
+
+class FunctionCastJsonAsJson : public IFunction
+{
+public:
+    static constexpr auto name = "cast_json_as_json";
+    static FunctionPtr create(const Context &) { return std::make_shared<FunctionCastJsonAsJson>(); }
+
+    String getName() const override { return name; }
+
+    size_t getNumberOfArguments() const override { return 1; }
+
+    bool useDefaultImplementationForNulls() const override { return false; }
+    bool useDefaultImplementationForConstants() const override { return false; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override { return arguments[0]; }
+
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result) const override
+    {
+        auto from = block.getByPosition(arguments[0]).column;
+        block.getByPosition(result).column = std::move(from);
+    }
+};
+
+class FunctionCastRealAsJson : public IFunction
+{
+public:
+    static constexpr auto name = "cast_real_as_json";
+    static FunctionPtr create(const Context &) { return std::make_shared<FunctionCastRealAsJson>(); }
+
+    String getName() const override { return name; }
+
+    size_t getNumberOfArguments() const override { return 1; }
+
+    bool useDefaultImplementationForNulls() const override { return true; }
+    bool useDefaultImplementationForConstants() const override { return true; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        if unlikely (!arguments[0]->isFloatingPoint())
+            throw Exception(
+                fmt::format("Illegal type {} of argument of function {}", arguments[0]->getName(), getName()),
+                ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+        return std::make_shared<DataTypeString>();
+    }
+
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result) const override
+    {
+        auto col_to = ColumnString::create();
+        auto & data_to = col_to->getChars();
+        JsonBinary::JsonBinaryWriteBuffer write_buffer(data_to);
+        auto & offsets_to = col_to->getOffsets();
+        auto rows = block.rows();
+        offsets_to.resize(rows);
```

Comment:
maybe `data_to` can also reserve enough capacity since the length is fixed and we already knows the row_size.

Reply:
+1